apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs

```diff
@@ -91,22 +141,245 @@ public sealed class PooledSession : IDisposable
         private int _disposed;
         private int _committedOrRolledBack;
 
-        private PooledSession(SessionPool.ISessionPool pool, SessionName sessionName, ByteString transactionId, ModeOneofCase transactionMode, Timestamp readTimestamp, DateTime evictionTime, long refreshTicks)
+        private PooledSession(SessionPool.ISessionPool pool, SessionName sessionName, ByteString transactionId, TransactionOptions transactionOptions, bool singleUseTransaction, Timestamp readTimestamp, DateTime evictionTime, long refreshTicks)
         {
-            GaxPreconditions.CheckArgument(
-                (transactionId == null) == (transactionMode == ModeOneofCase.None),
-                nameof(transactionMode),
-                "Transaction mode and ID don't match.");
             _pool = pool;
-            SessionName = GaxPreconditions.CheckNotNull(sessionName, nameof(sessionName));
-            TransactionId = transactionId;
-            TransactionMode = transactionMode;
-            ReadTimestamp = readTimestamp;
-            _session = new Session { SessionName = SessionName };
+            Session = new Session
+            {
+                SessionName = GaxPreconditions.CheckNotNull(sessionName, nameof(sessionName))
+            };
+            TransactionOptions = transactionOptions?.Clone() ?? new TransactionOptions();
+
+            GaxPreconditions.CheckArgument(
+                TransactionOptions.ModeCase == ModeOneofCase.ReadOnly || !singleUseTransaction,
+                nameof(singleUseTransaction),
+                "Single use transactions are only supported for read-only transactions.");
+            GaxPreconditions.CheckArgument(
+                transactionId is null || TransactionOptions.ModeCase != ModeOneofCase.None,
+                nameof(transactionOptions),
+                "No transaction options were specified for the given transasaction ID.");
+            GaxPreconditions.CheckArgument(
+                readTimestamp is null || transactionId is not null,
+                nameof(readTimestamp),
+                "A read timestamp can only be specified if a transaction ID is also being specified.");
+
+            SingleUseTransaction = singleUseTransaction;
+            if (transactionId is not null)
+            {
+                // We don't need to lock here, this is the constructor.
+                _transaction = new Transaction
```

Comment:
I think this is so that users doing partitions need only remember the transaction ID and not the rest of the transaction fields? But possibly we can handle that elsewhere and here just receive the whole Transaction. I'll leave a TODO to explore that when I look at detached transactions.